library/system/io.lean

```diff
@@ -68,6 +91,14 @@ interface.iterate e α a f
 def forever {e} (a : io_core e unit) : io_core e unit :=
 iterate () $ λ _, a >> return (some ())
 
+def catch {e₁ e₂ α} (a : io_core e₁ α) (b : e₁ → io_core e₂ α) : io_core e₂ α :=
+interface.catch e₁ e₂ α a b
+
+instance : alternative io :=
+{ interface.monad _ with
+  orelse := λ _ a b, catch a (λ _, b),
+  failure := λ _, io.fail "failure" }
+
```

Comment:
@gebner @kha I'm not sure if this is high priority, but these laws are essential for anybody trying to prove things about code using these type classes. 

BTW, I also tried a quick experiment with the state monad this week. The state monad as-is is quite bad for verification because there is no easy way to state properties about the state encapsulated in the monad. I defined a monad inspired by HTT (Hoare Type Theory). Here is an example: https://gist.github.com/leodemoura/15e4f19f325139b17575522bce80dd80
This approach will only be convenient to use after we add tactics for automating lemmas such as `init_mem_inv`.